<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>vcf2others • vcf2others</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="vcf2others">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">vcf2others</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/vcf2others.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>vcf2others</h1>
                        <h4 data-toc-skip class="author">Tomas
Hrbek</h4>
            
            <h4 data-toc-skip class="date">11/26/2023</h4>
      

      <div class="d-none name"><code>vcf2others.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>This package needs to be installed from GitHub.
<code>devtools::install_github("legalLab/vcf2others")</code></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package has two main functionalities. First, it is used for
filtering, subsetting, merging and otherwise wrangling VCF files.
Second, it is used for converting the VCF file to other population
genetic and phylogenetic formats for downstream analyses.</p>
</div>
<div class="section level2">
<h2 id="how-to-use-the-functions-of-this-package">How to use the functions of this package<a class="anchor" aria-label="anchor" href="#how-to-use-the-functions-of-this-package"></a>
</h2>
<p>Following are examples of the usage of the functions of this
package.</p>
<div class="section level3">
<h3 id="load-vcf-and-associated-files-of-individuals-and-groups">Load VCF and associated files of individuals and groups<a class="anchor" aria-label="anchor" href="#load-vcf-and-associated-files-of-individuals-and-groups"></a>
</h3>
<p>First we need to load our files. In addition to the VCF, we also need
to have a white/black list of individuals (read from a file of one
column with an ‘id’ header), a strata file associating individuals with
groups (read from a file of two tab-separated columns with an ‘id’ and
‘pop’ headers), and a file of groups (read from a file of one column
with a ‘pop’ header). These auxilary files are used for
keeping/excluding individuals and groups of individuals during filtering
and during conversion to other data formats. The strata file is also
necessary since many of the population genetic formats require
information on grouping of individuals.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/knausb/vcfR" class="external-link">vcfR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    *****       ***   vcfR   ***       *****</span></span>
<span><span class="co">#&gt;    This is vcfR 1.15.0 </span></span>
<span><span class="co">#&gt;      browseVignettes('vcfR') # Documentation</span></span>
<span><span class="co">#&gt;      citation('vcfR') # Citation</span></span>
<span><span class="co">#&gt;    *****       *****      *****       *****</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://legallab.github.io/vcf2others/">vcf2others</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set path to example files and project name</span></span>
<span><span class="va">data_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"vcf2others"</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">project</span> <span class="op">&lt;-</span> <span class="st">"crocs_"</span></span>
<span><span class="va">postfix</span> <span class="op">&lt;-</span> <span class="st">"discosnp_sub"</span></span>
<span></span>
<span><span class="co"># load vcf</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vcfR/man/io_vcfR.html" class="external-link">read.vcfR</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">".vcf.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scanning file to determine attributes.</span></span>
<span><span class="co">#&gt; File attributes:</span></span>
<span><span class="co">#&gt;   meta lines: 22</span></span>
<span><span class="co">#&gt;   header_line: 23</span></span>
<span><span class="co">#&gt;   variant count: 10000</span></span>
<span><span class="co">#&gt;   column count: 26</span></span>
<span><span class="co">#&gt; Meta line 22 read in.</span></span>
<span><span class="co">#&gt; All meta lines processed.</span></span>
<span><span class="co">#&gt; gt matrix initialized.</span></span>
<span><span class="co">#&gt; Character matrix gt created.</span></span>
<span><span class="co">#&gt;   Character matrix gt rows: 10000</span></span>
<span><span class="co">#&gt;   Character matrix gt cols: 26</span></span>
<span><span class="co">#&gt;   skip: 0</span></span>
<span><span class="co">#&gt;   nrows: 10000</span></span>
<span><span class="co">#&gt;   row_num: 0</span></span>
<span><span class="co">#&gt; Processed variant 1000Processed variant 2000Processed variant 3000Processed variant 4000Processed variant 5000Processed variant 6000Processed variant 7000Processed variant 8000Processed variant 9000Processed variant 10000Processed variant: 10000</span></span>
<span><span class="co">#&gt; All variants processed</span></span>
<span></span>
<span><span class="co"># read individuals to include</span></span>
<span><span class="va">indivs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">"indivs_b"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check if all indivs are in vcf_names</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">indivs</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">@</span><span class="va">gt</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Some individuals in list not in VCF"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get groups and group info on samples in vcf</span></span>
<span><span class="co"># default file names for strata is "strata" and for groups "groups"</span></span>
<span><span class="va">project_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_vcf_group_info.html">get_vcf_group_info</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">data_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(id)`</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="assessment-of-missing-data">Assessment of Missing Data<a class="anchor" aria-label="anchor" href="#assessment-of-missing-data"></a>
</h3>
<p>The function <code><a href="../reference/assess_vcf_missing_data.html">assess_vcf_missing_data()</a></code> is used to
generate a table and graph of missing data for each sample in a VCF. It
is useful for visualizing data before and after filtering, and evaluate
the effect of filtering. The function accepts parameters which form part
of the name of the input file. The idea is that the VCF name contains
information on the name of the project (project), how the VCF was
extracted (postfix), and how it was filtered (fltr). The “postfix” and
“fltr” can be left blank if the VCF name does not contain this
information. Samples are ordered and colored by group assignment.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define results path</span></span>
<span><span class="va">res_path</span> <span class="op">&lt;-</span> <span class="va">data_path</span></span>
<span><span class="co"># species for plot title</span></span>
<span><span class="va">species</span> <span class="op">&lt;-</span> <span class="st">"Paleosuchus/Caiman"</span></span>
<span><span class="co"># filter - filter parameters in file name</span></span>
<span><span class="va">fltr</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span></span>
<span><span class="fu"><a href="../reference/assess_vcf_missing_data.html">assess_vcf_missing_data</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">data_path</span>, <span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="va">fltr</span>, <span class="va">species</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(id)`</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="basic-stats-for-individuals-in-vcf">Basic stats for individuals in VCF<a class="anchor" aria-label="anchor" href="#basic-stats-for-individuals-in-vcf"></a>
</h3>
<p>The function <code><a href="../reference/vcf_stats.html">vcf_stats()</a></code> is used to generate a table
with basic stats for each sample in the VCF. These stats include average
read depth, heterozygosity, number of heterozygotes, homozygotes,
reference homozygotes and alternate homogygotes, percent missing SNPs,
total missing SNPs and total SNPs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get a table of basic sample stats</span></span>
<span></span>
<span><span class="fu"><a href="../reference/vcf_stats.html">vcf_stats</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">res_path</span>, <span class="va">project</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filtering-subsetting-merging-and-otherwise-wrangling-vcf-files">Filtering, subsetting, merging and otherwise wrangling VCF
files<a class="anchor" aria-label="anchor" href="#filtering-subsetting-merging-and-otherwise-wrangling-vcf-files"></a>
</h3>
<p>All functions can be used directly on a VCF, but many of them are
called by the other functions of this package. For example
<code><a href="../reference/vcf_filter_invariant.html">vcf_filter_invariant()</a></code> will remove invariant SNPs from a
VCF, however, VCF by definition should not have invariant SNPs. However,
if a VCF is subsetted or certain individuals are removed, SNPs can
become invariant and need to be removed.</p>
<div class="section level4">
<h4 id="subsetting">Subsetting<a class="anchor" aria-label="anchor" href="#subsetting"></a>
</h4>
<p>The functions <code><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs()</a></code>,
<code><a href="../reference/vcf_extract_pops.html">vcf_extract_pops()</a></code> and <code><a href="../reference/vcf_sub_loci.html">vcf_sub_loci()</a></code> are used
to subset a VCF by specific individuals, a group (all individuals of a
group/population), or a random subset of loci, respectively. Subsetting
individuals and groups can be done either as a whitelist (keep the
listed individuals/groups) or blacklist (remove the listed
individuals/groups).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset VCF by individuals (default whitelist = TRUE, remove invariants = TRUE)</span></span>
<span><span class="va">indivs1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTGA_H4635"</span>, <span class="st">"CTGA_H4667"</span><span class="op">)</span></span>
<span><span class="va">vcf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">indivs1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset VCF by groups (default whitelist = TRUE, remove invariants = TRUE)</span></span>
<span><span class="va">groups1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"paleosuchus_1"</span>, <span class="st">"paleosuchus_2"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">vcf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_pops.html">vcf_extract_pops</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">groups1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset VCF by n random loci (default n_loci = 1000)</span></span>
<span><span class="va">vcf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_sub_loci.html">vcf_sub_loci</a></span><span class="op">(</span><span class="va">vcf</span>, n_loci <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filtering">Filtering<a class="anchor" aria-label="anchor" href="#filtering"></a>
</h4>
<p>Filtering consists of removing individuals and loci based on specific
properties. The function <code><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs()</a></code>
removes individuals above a % missing SNP threshold. The other functions
filter on properties of individuals SNPs across individuals. The
function <code><a href="../reference/vcf_filter_missingness.html">vcf_filter_missingness()</a></code> removes SNPs above a %
missing threshold, <code><a href="../reference/vcf_filter_quality.html">vcf_filter_quality()</a></code> removes SNPs below
a quality threshold, <code><a href="../reference/vcf_filter_maf.html">vcf_filter_maf()</a></code> removes SNPs with
minor allele frequency below a % threshold,
<code><a href="../reference/vcf_filter_hets.html">vcf_filter_hets()</a></code> removes SNPs with heterozigosity above a
% threshold, <code><a href="../reference/vcf_filter_rank.html">vcf_filter_rank()</a></code> removes SNPs with rank below
a % threshold (rank is specific to VCFs generated by DiscoSnpRad and
measures the likelihood of observed SNP variation being due to
paralogs), and <code><a href="../reference/vcf_filter_coverage.html">vcf_filter_coverage()</a></code> removes SNPs below a
specific coverage (makes them missing in the individual). Finally the
<code><a href="../reference/vcf_filter_invariant.html">vcf_filter_invariant()</a></code> removes invariant SNPs (SNPs may
become invariant after subsetting and filtering). The functions
<code><a href="../reference/vcf_filter_oneSNP.html">vcf_filter_oneSNP()</a></code> and <code><a href="../reference/vcf_filter_multiSNP.html">vcf_filter_multiSNP()</a></code>
filter a VCF to remove linked SNPs (retain only unliked SNPs) or retain
only linked SNPs, respectively. In general, all analyses assume that
SNPs are unlinked, however, some analyses such as FineRadStructure work
with linked SNPs.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter VCF for analyses (unlinked SNPs)</span></span>
<span><span class="va">vcf_oneSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">indivs</span>, whitelist <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs</a></span><span class="op">(</span><span class="fl">.8</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_rank.html">vcf_filter_rank</a></span><span class="op">(</span><span class="fl">.5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_maf.html">vcf_filter_maf</a></span><span class="op">(</span><span class="fl">.03</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_coverage.html">vcf_filter_coverage</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_oneSNP.html">vcf_filter_oneSNP</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missingness.html">vcf_filter_missingness</a></span><span class="op">(</span><span class="fl">.2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs</a></span><span class="op">(</span><span class="fl">.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; removed samples are: CTGA_H4663 </span></span>
<span><span class="co">#&gt;  removed samples are: CTGA_H4668 </span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 54.91 %"</span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 7.66 %"</span></span>
<span><span class="co">#&gt; removed samples are: CTGA_H4667 </span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 5.74 %"</span></span>
<span></span>
<span><span class="co"># filter VCF for analyses (linked SNPs)</span></span>
<span><span class="va">vcf_multiSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">indivs</span>, whitelist <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs</a></span><span class="op">(</span><span class="fl">.8</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_rank.html">vcf_filter_rank</a></span><span class="op">(</span><span class="fl">.5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_maf.html">vcf_filter_maf</a></span><span class="op">(</span><span class="fl">.03</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_coverage.html">vcf_filter_coverage</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_multiSNP.html">vcf_filter_multiSNP</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missingness.html">vcf_filter_missingness</a></span><span class="op">(</span><span class="fl">.2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs</a></span><span class="op">(</span><span class="fl">.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; removed samples are: CTGA_H4663 </span></span>
<span><span class="co">#&gt;  removed samples are: CTGA_H4668 </span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 54.91 %"</span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 6.67 %"</span></span>
<span><span class="co">#&gt; removed samples are: CTGA_H4667 </span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 4.57 %"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extracting-merging-and-adding">Extracting, merging and adding<a class="anchor" aria-label="anchor" href="#extracting-merging-and-adding"></a>
</h4>
<p>The functions <code><a href="../reference/vcf_merge.html">vcf_merge()</a></code>,
<code><a href="../reference/vcf_add_indivs.html">vcf_add_indivs()</a></code>, <code><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs()</a></code> and
<code><a href="../reference/vcf_extract_pops.html">vcf_extract_pops()</a></code> are used to merge two VCF files, add
individuals from a second VCF into a first VCF, and extract
individuals/groups from a VCF saving it into a separate VCF. The
<code><a href="../reference/vcf_add_indivs.html">vcf_add_indivs()</a></code>, <code><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs()</a></code> and
<code><a href="../reference/vcf_extract_pops.html">vcf_extract_pops()</a></code> functions were designed in mind to work
with outgroups. Because of phylogenetic effects, outgroups have
relatively more missing data than ingroups, and so often are removed
during filtering. The solution is to extract the outgroup taxa with
<code><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs()</a></code> or <code><a href="../reference/vcf_extract_pops.html">vcf_extract_pops()</a></code>,
filter the ingroup VCF, and add the outgroup taxa with
<code><a href="../reference/vcf_add_indivs.html">vcf_add_indivs()</a></code>. It is important to NOT filter invariant
loci during the filtering of the ingroup, and only filter them
<code><a href="../reference/vcf_filter_invariant.html">vcf_filter_invariant()</a></code> after adding back the outgroup. The
functions <code><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs()</a></code>,
<code><a href="../reference/vcf_extract_pops.html">vcf_extract_pops()</a></code>,
<code><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs()</a></code> and
<code><a href="../reference/vcf_filter_coverage.html">vcf_filter_coverage()</a></code> by default filter invariants, but
this can be turned off using the “f_invar = FALSE” flag.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract individuals from VCF (keep all loci)</span></span>
<span><span class="va">indivs1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTGA_H4635"</span>, <span class="st">"CTGA_H4637"</span>, <span class="st">"CTGA_H4643"</span>, <span class="st">"CTGA_H4645"</span>, <span class="st">"CTGA_H4647"</span><span class="op">)</span></span>
<span><span class="va">vcf_outgrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">indivs1</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vcf_ingrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_indivs.html">vcf_extract_indivs</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">indivs1</span>, whitelist <span class="op">=</span> <span class="cn">FALSE</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract groups of individuals from VCF (keep all loci)</span></span>
<span><span class="va">groups1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="st">"caiman"</span><span class="op">)</span></span>
<span><span class="va">vcf_outgrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_pops.html">vcf_extract_pops</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">groups1</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vcf_ingrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_pops.html">vcf_extract_pops</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">groups1</span>, whitelist <span class="op">=</span> <span class="cn">FALSE</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># merge vcf_outgrp into vcf_ingrp</span></span>
<span><span class="va">vcf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_merge.html">vcf_merge</a></span><span class="op">(</span><span class="va">vcf_ingrp</span>, <span class="va">vcf_outgrp</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(FORMAT, id)`</span></span>
<span></span>
<span><span class="co"># add individuals from vcf_outgrp into vcf_ingrp (whitelist = TRUE is default)</span></span>
<span><span class="va">indivs1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTGA_H4637"</span>, <span class="st">"CTGA_H4643"</span>, <span class="st">"CTGA_H4645"</span><span class="op">)</span></span>
<span><span class="va">vcf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_add_indivs.html">vcf_add_indivs</a></span><span class="op">(</span><span class="va">vcf_ingrp</span>, <span class="va">vcf_outgrp</span>, <span class="va">indivs1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(FORMAT, id)`</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filtering-with-outgroup">Filtering with outgroup<a class="anchor" aria-label="anchor" href="#filtering-with-outgroup"></a>
</h4>
<p>Outgroups, which tend to have many fewer individuals than ingroups,
has more missing data than ingroups due to phylogenetic effects.
Filtering the entire dataset often results in removal of the outgroup
taxa when <code><a href="../reference/vcf_filter_invariant.html">vcf_filter_invariant()</a></code> is used during filtering
to remove individuals with % missing data above some threshold.
Therefore, a better strategy may be to extract the outgroup, filter the
ingroup, add the outgroup, and remove any invariant SNPs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract outgroup from VCF (keep all loci)</span></span>
<span><span class="va">groups1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="st">"caiman"</span><span class="op">)</span></span>
<span><span class="va">vcf_outgrp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_pops.html">vcf_extract_pops</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">groups1</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter ingroup VCF for analyses then add outgroups</span></span>
<span><span class="va">vcf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vcf_extract_pops.html">vcf_extract_pops</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">groups1</span>, whitelist <span class="op">=</span> <span class="cn">FALSE</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs</a></span><span class="op">(</span><span class="fl">.8</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_rank.html">vcf_filter_rank</a></span><span class="op">(</span><span class="fl">.5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_maf.html">vcf_filter_maf</a></span><span class="op">(</span><span class="fl">.03</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_coverage.html">vcf_filter_coverage</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_oneSNP.html">vcf_filter_oneSNP</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missingness.html">vcf_filter_missingness</a></span><span class="op">(</span><span class="fl">.2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_missing_indivs.html">vcf_filter_missing_indivs</a></span><span class="op">(</span><span class="fl">.3</span>, f_invar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_merge.html">vcf_merge</a></span><span class="op">(</span><span class="va">vcf_outgrp</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/vcf_filter_invariant.html">vcf_filter_invariant</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; removed samples are: CTGA_H4663 </span></span>
<span><span class="co">#&gt;  removed samples are: CTGA_H4668 </span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 57.52 %"</span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is 3.33 %"</span></span>
<span><span class="co">#&gt; removed samples are: CTGA_H4667 </span></span>
<span><span class="co">#&gt; [1] "final % missing data in VCF is NaN %"</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(FORMAT, id)`</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="converting-a-vcf-file-to-other-population-genetic-and-phylogenetic-formats">Converting a VCF file to other population genetic and phylogenetic
formats<a class="anchor" aria-label="anchor" href="#converting-a-vcf-file-to-other-population-genetic-and-phylogenetic-formats"></a>
</h3>
<p>All the following functions will take a VCF, groups and indiv_group
(individuals to groups relationship), and convert the VCF to other
populations genetic and phylogenetic formats, exporting/writing a file
of this format. By default all functions will include missing data. If
that is not desired, only 100% complete data matrices can be exported
using the “inc_missing = FALSE” flag. The function
<code><a href="../reference/vcf2genlight.html">vcf2genlight()</a></code> automatically returns a genlight object and
optionally can also export/write a file of this format to the working
directory. Generally the <code><a href="../reference/vcf2genlight.html">vcf2genlight()</a></code> function is called
within a script using functions of the <code>adegenet</code> package
rather than importing the genlight object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_path</span> <span class="op">&lt;-</span> <span class="va">data_path</span></span>
<span><span class="va">project</span> <span class="op">&lt;-</span> <span class="st">"trigonatus_"</span></span>
<span><span class="va">postfix</span> <span class="op">&lt;-</span> <span class="st">"discosnp_sub"</span></span>
<span></span>
<span><span class="co">##########</span></span>
<span><span class="co"># datasets for analyses with unlinked SNPs</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="va">vcf_oneSNP</span></span>
<span><span class="co"># get groups and group info on samples in vcf</span></span>
<span><span class="va">project_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_vcf_group_info.html">get_vcf_group_info</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">data_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(id)`</span></span>
<span></span>
<span><span class="co">##########</span></span>
<span><span class="co"># export data formats</span></span>
<span><span class="co"># migrate-n https://peterbeerli.com/migrate-html5/</span></span>
<span><span class="fu"><a href="../reference/vcf2migrate.html">vcf2migrate</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'_migrate.txt'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># arlequin http://cmpg.unibe.ch/software/arlequin35/</span></span>
<span><span class="fu"><a href="../reference/vcf2arlequin.html">vcf2arlequin</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.arp'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># structure https://web.stanford.edu/group/pritchardlab/structure.html</span></span>
<span><span class="fu"><a href="../reference/vcf2structure.html">vcf2structure</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.str'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># faststucture http://rajanil.github.io/fastStructure/</span></span>
<span><span class="fu"><a href="../reference/vcf2structure.html">vcf2structure</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.fstr'</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span></span>
<span><span class="co"># genepop https://gitlab.mbb.univ-montp2.fr/francois/genepop</span></span>
<span><span class="fu"><a href="../reference/vcf2genepop.html">vcf2genepop</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.gen'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># smartsnp https://github.com/ChristianHuber/smartsnp</span></span>
<span><span class="fu"><a href="../reference/vcf2smartsnp.html">vcf2smartsnp</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.smartsnp'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># eigenstrat https://github.com/DReichLab/EIG/tree/master</span></span>
<span><span class="fu"><a href="../reference/vcf2eigenstrat.html">vcf2eigenstrat</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'_eigenstrat'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># bayescan https://github.com/mfoll/BayeScan</span></span>
<span><span class="fu"><a href="../reference/vcf2bayescan.html">vcf2bayescan</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.bayescan'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># treemix https://bitbucket.org/nygcresearch/treemix/wiki/Home</span></span>
<span><span class="fu"><a href="../reference/vcf2treemix.html">vcf2treemix</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.treemix'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># apparent https://github.com/halelab/apparent/tree/master</span></span>
<span><span class="fu"><a href="../reference/vcf2apparent.html">vcf2apparent</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.apparent'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># related https://github.com/timothyfrasier/related</span></span>
<span><span class="fu"><a href="../reference/vcf2related.html">vcf2related</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.related'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># dataframe with population membership in the last column</span></span>
<span><span class="fu"><a href="../reference/vcf2df.html">vcf2df</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.df'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># snapp https://www.beast2.org/snapp/</span></span>
<span><span class="fu"><a href="../reference/vcf2snapp.html">vcf2snapp</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'_snapp.nex'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># nexus - only SNPs, meant for SDVq analyses https://www.asc.ohio-state.edu/kubatko.2/software/SVDquartets/</span></span>
<span><span class="fu"><a href="../reference/vcf2nexus.html">vcf2nexus</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'_sdvq.nex'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># fasta https://www.ncbi.nlm.nih.gov/genbank/fastaformat/</span></span>
<span><span class="fu"><a href="../reference/vcf2fasta.html">vcf2fasta</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.fna'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##########</span></span>
<span><span class="co"># datasets for analyses with linked SNPs</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="va">vcf_multiSNP</span></span>
<span><span class="co"># get groups and group info on samples in vcf</span></span>
<span><span class="va">project_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_vcf_group_info.html">get_vcf_group_info</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">data_path</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(id)`</span></span>
<span></span>
<span><span class="co">##########</span></span>
<span><span class="co"># export data formats</span></span>
<span><span class="co"># fineRadStructure - expects VCF of linked SNPs</span></span>
<span><span class="fu"><a href="../reference/vcf2fineRadStructure.html">vcf2fineRadStructure</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">indiv_group</span>, <span class="va">project_info</span><span class="op">$</span><span class="va">groups</span>, out_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">res_path</span>, <span class="va">project</span>, <span class="va">postfix</span>, <span class="st">'.finerad'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tomas Hrbek.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
