#' @title vcf2genepop
#' @description converts vcfR format data to Genepop infile
#' @description includes population names that can be ready by read_genepop {}
#' @description in part based on vcfR2migrate function (vcfR package)
#' @author Tomas Hrbek January 2021
#'
#' @param vcf -> vcfR object
#' @param ind_pop -> population assignment of individuals in vcf (factor)
#' @param keep_pop -> population(s) of interest to include in Genepop infile (factor)
#' @param inc_missing -> include missing data (logical)
#' @param out_file -> name of file to output (Genepop infile)
#' @export
#' @return nothing
#'
#' @details
#' This function converts the vcfR object to a Genepop formatted input file
#' This function labels populations. To read labeled populations use "read_genepop" function
#' The function will remove indels, and multiallelic loci, and optionally loci with missing data
#' 01, 02, 03, 04 is 'A', 'C', 'G', 'T'
#'
#' @examples
#' vcf2genepop(vcf = my_vcf, ind_pop = ind_pop, keep_pop = keepers, inc_missing = TRUE, out_file = "Genepop_infile.gen")
#' vcf2genepop(my_vcf, ind_pop, keepers, out_file = "Genepop_infile.gen")
#' vcf2genepop(my_vcf, ind_pop, keepers)
#'

vcf2genepop <- function (vcf, ind_pop, keep_pop, inc_missing = TRUE,
                         out_file = "genepop_infile.txt")
{
  if (class(vcf) != "vcfR") {
    stop(paste("Expecting an object of class vcfR, received a",
               class(vcf), "instead"))
  }
  if (class(ind_pop) != "factor" | class(keep_pop) != "factor") {
    stop(paste("Expecting population vector, received a",
               class(ind_pop), "and", class(keep_pop), "instead"))
  }
  vcf <- vcfR::extract.indels(vcf, return.indels = F)
  vcf <- vcf[vcfR::is.biallelic(vcf), ]
  if (inc_missing == FALSE) {
    gt <- vcfR::extract.gt(vcf, convertNA = T)
    vcf <- vcf[!rowSums((is.na(gt))), ]
  }
  vcf_list <- lapply(keep_pop, function(x) {
    vcf[, c(TRUE, x == ind_pop)]
  })
  names(vcf_list) <- keep_pop
  pop_list <- vector(mode = "list", length = length(vcf_list))
  names(pop_list) <- names(vcf_list)

  gt <- vcfR::extract.gt(vcf, return.alleles = F, convertNA = T)
  gt <- t(gt)

  write("Title = 'Generated by vcf2genpop.R'", file = out_file)
  suppressWarnings(utils::write.table(gt[0,], file = out_file, quote = FALSE, sep = ", ", col.names = TRUE, append = TRUE))

  for (i in 1:length(vcf_list)) {
    gt <- vcfR::extract.gt(vcf_list[[i]], return.alleles = T, convertNA = T) #convertNA not working here
    gt[gt == "."] <- "0000"
    gt[gt == "A/A" | gt == "A|A"] <- "0101"
    gt[gt == "A/C" | gt == "C/A" | gt == "A|C" | gt == "C|A"] <- "0102"
    gt[gt == "A/G" | gt == "G/A" | gt == "A|G" | gt == "G|A"] <- "0103"
    gt[gt == "A/T" | gt == "T/A" | gt == "A|T" | gt == "T|A"] <- "0104"
    gt[gt == "C/C" | gt == "C|C"] <- "0202"
    gt[gt == "C/G" | gt == "G/C" | gt == "C|G" | gt == "G|C"] <- "0203"
    gt[gt == "C/T" | gt == "T/C" | gt == "C|T" | gt == "T|C"] <- "0204"
    gt[gt == "G/G" | gt == "G|G"] <- "0303"
    gt[gt == "G/T" | gt == "T/G" | gt == "G|T" | gt == "T|G"] <- "0304"
    gt[gt == "T/T" | gt == "T|T"] <- "0404"
    gt <- t(gt)
    write(paste("pop ", names(pop_list)[i], sep = ""), file = out_file, append = TRUE)
    utils::write.table(cbind(sep = ',', gt), file = out_file, quote = FALSE, sep = " ", col.names = FALSE, append = TRUE)
  }

  return(invisible(NULL))
}
